### PR TITLE
Kafka07 with Thrift Decoder

### PR DESCRIPTION
**Pull request by xiaoyao1991**

### CODE REVIEW

**Comment by huijunw** [comment]
good job

**Comment by billonahill** [comment]
This patch includes a large amount of inline changes to the code on the master. If we can't contribute these changes back upstream, this could effectively fork our distro if merging from the master gets too nasty in the future. Is it possible to replace many of these in-line changes with 1 liner calls that delegate to methods at the bottom of the class, or better yet to classes (utilities or subclasses) in the `c.f.p.twitter`package? The goal would be to make it easy to merge from master later without conflict or with one that is easy to resolve.

**Comment by xiaoyao1991** [comment]
@billonahill I see your point. I'm thinking of making the kafka07 another plugin, so that it lives side by side with the upstream kafka08 plugin. But the cons of that is code duplicate, though any new changes from upstream may not be beneficial us anyway.

**Comment by billonahill** [comment]
@xiaoyao1991 I think the kafka07 connector might be the best approach. Then the challenge would be how to do these code reviews. You could commit a kafka07 connector that is the same as the current kafka connector, then we can submit patches on that. Also it would be helpful for reviews if we could break this large patch into a few smaller PRs for ease of consumption.

File: presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaRecordSet.java [hunk 1]
```diff
@@ -158,50 +169,45 @@ public boolean advanceNextPosition()
         {
             while (true) {
                 if (cursorOffset >= split.getEnd()) {
-                    return endOfData(); // Split end is exclusive.
+                    return endOfData(1); // Split end is exclusive.
```
Comment (herval) [inline presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaRecordSet.java]: what does this comment mean?

File: presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaRecordSet.java [hunk 2]
```diff
@@ -158,50 +169,45 @@ public boolean advanceNextPosition()
...
-                    MessageAndOffset currentMessageAndOffset = messageAndOffsetIterator.next();
-                    long messageOffset = currentMessageAndOffset.offset();
 
-                    if (messageOffset >= split.getEnd()) {
-                        return endOfData(); // Past our split end. Bail.
+                try {
+                    // Create a fetch request
+                    openFetchRequest();
+                    if (cursorOffset >= split.getEnd()) {
+                        return endOfData(2); // Split end is exclusive.
```
Comment (herval) [inline presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaRecordSet.java]: ditto

File: presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaSimpleConsumerManager.java [hunk 3]
```diff
@@ -94,12 +93,11 @@ public SimpleConsumer getConsumer(HostAddress host)
...
                 throws Exception
         {
-            log.info("Creating new Consumer for %s", host);
+            log.debug("Creating new Consumer for %s", host);
             return new SimpleConsumer(host.getHostText(),
                     host.getPort(),
                     connectTimeoutMillis,
-                    bufferSizeBytes,
-                    format("presto-kafka-%s-%s", connectorId, nodeManager.getCurrentNode().getNodeIdentifier()));
+                    bufferSizeBytes);
```
Comment (herval) [inline presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaSimpleConsumerManager.java]: how come you don't need the identifier bit anymore (the param you removed)?

File: presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaSplitManager.java [hunk 4]
```diff
@@ -69,76 +61,97 @@ public KafkaSplitManager(
...
-        OffsetResponse offsetResponse = consumer.getOffsetsBefore(offsetRequest);
-
-        if (offsetResponse.hasError()) {
-            short errorCode = offsetResponse.errorCode(topicName, partitionId);
-            log.warn("Offset response has error: %d", errorCode);
-            throw new PrestoException(KAFKA_SPLIT_ERROR, "could not fetch data from Kafka, error code is '" + errorCode + "'");
+        // startTs: start timestamp, or -2/null as earliest
+        // endTs: end timestamp, or -1/null as latest
+        startTs = startTs == null ? OffsetRequest.EarliestTime() : startTs;
+        endTs = endTs == null ? OffsetRequest.LatestTime() : endTs;
```
Comment (herval) [inline presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaSplitManager.java]: will these ever be null? The attribute on `layoutHandle` is already checked for nullability - you could perhaps just use `long` here?

File: presto-kafka/src/test/java/com/facebook/presto/kafka/util/EmbeddedKafka.java [hunk 5]
```diff
@@ -116,26 +116,15 @@ public void createTopics(String... topics)
...
-        ZkClient zkClient = new ZkClient(getZookeeperConnectString(), 30_000, 30_000, ZKStringSerializer$.MODULE$);
-        try {
-            for (String topic : topics) {
-                AdminUtils.createTopic(zkClient, topic, partitions, replication, topicProperties);
-            }
-        }
-        finally {
-            zkClient.close();
-        }
+        // noop
```
Comment (herval) [inline presto-kafka/src/test/java/com/facebook/presto/kafka/util/EmbeddedKafka.java]: why?

File: presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaRecordSet.java (hunk 1 above)
Comment (xiaoyao1991) [inline presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaRecordSet.java]: It means that the offset has reached or gone beyond the end of this split. A split in Kafka sense is just a range specified by 2 message offsets.

File: presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaSimpleConsumerManager.java (hunk 3 above)
Comment (xiaoyao1991) [inline presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaSimpleConsumerManager.java]: cuz in kafka07, it doesn't need that info...

File: presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaSplitManager.java (hunk 4 above)
Comment (xiaoyao1991) [inline presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaSplitManager.java]: good point.

File: presto-kafka/src/test/java/com/facebook/presto/kafka/util/EmbeddedKafka.java (hunk 5 above)
Comment (xiaoyao1991) [inline presto-kafka/src/test/java/com/facebook/presto/kafka/util/EmbeddedKafka.java]: kafka07 zk api doesn't support those methods I removed. Plus, the topic is automatically created when a message is produced to it.

**Comment by xiaoyao1991** [comment]
Closing, in lieu of #95  #96

**Pull request closed without merging**